sdl_viewer/src/lib.rs

```diff
@@ -38,4 +42,277 @@ pub mod color;
 pub mod graphic;
 pub mod node_drawer;
 
-pub use camera::Camera;
+use box_drawer::BoxDrawer;
+use camera::Camera;
+use color::YELLOW;
+use node_drawer::{NodeDrawer, NodeViewContainer};
+use point_viewer::octree::{self, Octree};
+use sdl2::event::{Event, WindowEvent};
+use sdl2::keyboard::Scancode;
+use sdl2::video::GLProfile;
+use std::cmp;
+use std::collections::HashMap;
+use std::error::Error;
+use std::sync::Arc;
+
+type OctreeFactoryCallback = fn(&String) -> Result<Box<Octree>, Box<Error>>;
+
+pub struct SdlViewer {
+    octree_factory_functions: HashMap<String, OctreeFactoryCallback>,
+}
+
+impl SdlViewer {
+    pub fn new() -> Self {
+        SdlViewer {
+            octree_factory_functions: HashMap::new(),
+        }
+    }
+
+    pub fn register_prefix(mut self, prefix: String, f: OctreeFactoryCallback) -> SdlViewer {
+        self.octree_factory_functions.insert(prefix, f);
+        self
+    }
+
+    pub fn run(self) {
+        let matches = clap::App::new("sdl_viewer")
+            .args(&[
+                clap::Arg::with_name("octree")
+                    .help("Input path of the octree.")
+                    .index(1)
+                    .required(true),
+                clap::Arg::with_name("cache_size_mb")
+                    .help(
+                        "Maximum cache size in MB for octree nodes in GPU memory. \
+                         The default value is 2000 MB and the valid range is 1000 MB to 16000 MB.",
+                    )
+                    .required(false),
+            ])
+            .get_matches();
+
+        let octree_path = matches.value_of("octree").unwrap();
+
+        // Maximum number of MB for the octree node cache. The default is 2 GB
+        let cache_size_mb = matches
+            .value_of("cache_size_mb")
+            .unwrap_or("2000")
+            .parse()
+            .unwrap();
+
+        // Maximum number of MB for the octree node cache in range 1..16 GB. The default is 2 GB
+        let limit_cache_size_mb = cmp::max(1000, cmp::min(16_000, cache_size_mb));
+
+        // Assuming about 200 KB per octree node on average
+        let max_nodes_in_memory = limit_cache_size_mb * 5;
+
+        // call octree generation functions
+        let mut octree_opt: Option<Box<Octree>> = None;
+        for (prefix, octree_factory_function) in &self.octree_factory_functions {
+            if !octree_path.starts_with(prefix) {
+                continue;
+            }
+            let mut no_prefix_path = String::from(octree_path);
```

Comment:
could also use octree_path.to_string(). I have no preference really.

Reply:
Then again, this should work to and it will not clone: 
`let no_prefix = octree_path[prefix.len()..]`